lib/rules/semi.js

```diff
@@ -102,7 +120,13 @@ module.exports = function(context) {
 
         if (always) {
             if (!isSemicolon(lastToken)) {
-                report(node);
+                if (!exceptOneline) {
+                    report(node);
+                } else {
+                    if (!isOnelineException(lastToken)) {
+                        report(node);
+                    }
+                }
```

Comment:
I thought this can get simpler.

```js
if (!isSemicolon(lastToken) && (!exceptOneline || !isOnelineException(lastToken))) {
    report(node);
}
```